File: packages/relational-text/src/alignment.ts
/**
 * Types for cross-document and cross-layer alignment.
 * Derived from the pub.layers.alignment lexicons:
 *   - pub.layers.alignment.alignment
 *   - pub.layers.defs#alignmentLink
 *   - pub.layers.defs#objectRef
 */

import type { KnowledgeRef } from './knowledge.js'

/**
 * Composable object reference (pub.layers.defs#objectRef).
 * Re-exported here for alignment consumers.
 */
export interface ObjectRef {
  localId?: { value: string }
  recordRef?: string
  objectId?: { value: string }
  knowledgeRef?: KnowledgeRef
}

/**
 * A single link in an alignment between two parallel sequences.
 * Derived from pub.layers.defs#alignmentLink.
 */
export interface AlignmentLink {
  sourceIndices?: number[]
  targetIndices?: number[]
  /** Alignment confidence 0-1000. */
  confidence?: number
  label?: string
  knowledgeRefs?: KnowledgeRef[]
  features?: { entries: Array<{ key: string; value: string }> }
}

/**
 * An alignment between two parallel sequences.
 * Derived from pub.layers.alignment.alignment.
 */
export interface Alignment {
  expression?: string
  kindUri?: string
  kind: string
  subkindUri?: string
  subkind?: string
  source?: ObjectRef
  target?: ObjectRef
  sourceLang?: string
  targetLang?: string
  links: AlignmentLink[]
  knowledgeRefs?: KnowledgeRef[]
  features?: { entries: Array<{ key: string; value: string }> }
  createdAt?: string
}

File: crates/relationaltext-core/src/layers/alignment.rs
//! Alignment types for the Layers data model.
//!
//! Alignments map between parallel sequences (token-to-token across languages,
//! cross-tokenization mapping, interlinear glossing, etc.).

use serde::{Deserialize, Serialize};

use super::defs::{AlignmentLink, AnnotationMetadata, FeatureMap, ObjectRef};

/// An alignment between two parallel sequences.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Alignment {
    /// Reference to the source sequence (expression, tokenization, annotation layer, etc.).
    pub source_ref: ObjectRef,
    /// Reference to the target sequence.
    pub target_ref: ObjectRef,
    /// The alignment links.
    pub links: Vec<AlignmentLink>,
    /// Alignment kind (e.g., "word-alignment", "cross-tokenization", "interlinear").
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<AnnotationMetadata>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub features: Option<FeatureMap>,
    pub created_at: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::layers::defs::Uuid;

    #[test]
    fn alignment_roundtrip() {
        let alignment = Alignment {
            source_ref: ObjectRef {
                local_id: None,
                record_ref: Some("at://did:plc:abc/pub.layers.segmentation/seg-en".into()),
                object_id: Some(Uuid {
                    value: "tok-en".into(),
                }),
                knowledge_ref: None,
            },
            target_ref: ObjectRef {
                local_id: None,
                record_ref: Some("at://did:plc:abc/pub.layers.segmentation/seg-fr".into()),
                object_id: Some(Uuid {
                    value: "tok-fr".into(),
                }),
                knowledge_ref: None,
            },
            links: vec![
                AlignmentLink {
                    source_indices: Some(vec![0]),
                    target_indices: Some(vec![0, 1]),
                    confidence: Some(850),
                    label: None,
                    knowledge_refs: None,
                    features: None,
                },
                AlignmentLink {
                    source_indices: Some(vec![1]),
                    target_indices: Some(vec![2]),
                    confidence: None,
                    label: None,
                    knowledge_refs: None,
                    features: None,
                },
            ],
            kind: Some("word-alignment".into()),
            metadata: None,
            features: None,
            created_at: "2024-01-01T00:00:00Z".into(),
        };
        let json = serde_json::to_string(&alignment).unwrap();
        let back: Alignment = serde_json::from_str(&json).unwrap();
        assert_eq!(alignment, back);
    }
}
